bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeConverter.java

```diff
@@ -263,7 +264,44 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
         return convertibleTypes;
     }
 
-    private static boolean isConvertibleToRecordType(Object sourceValue, BRecordType targetType,
+    public static List<BType> getConvertibleTypesFromJson(Object value, BType targetType,
+                                                          List<TypeValuePair> unresolvedValues) {
+        List<BType> convertibleTypes = new ArrayList<>();
+
+        int targetTypeTag = targetType.getTag();
+
+        convertibleTypes.addAll(TypeConverter.getConvertibleTypes(value, targetType));
+
+        if (convertibleTypes.size() == 0) {
```

Comment:
`TypeConverter.getConvertibleTypes(value, targetType)` calls `isConvertibleToRecordType` with `isFromJson` set to `false` right?

What if `targetType` is a union of two records, say `Foo|Bar`, and the value is convertible to `Foo` even when  `isFromJson` is `false` in `isConvertibleToRecordType`, but it is only convertible to `Bar` when `isFromJson` is true?

So IINM, from L273 when we reach here `convertibleTypes` will only have `Foo` and we return that in L301. But if we actually check `isConvertibleToRecordType(value, (BRecordType) targetType, true, unresolvedValues)` it'll evaluate to `true` for `Bar`, and there'll actually be 2 convertible types?

Reply:
For example

```ballerina
import ballerina/io;

type Foo record {|
    int i;
    xml x;
|};

type Bar record {|
    int i;
    string x;
|};

type FooBar Foo|Bar;

public function main() {
    json j = {i: 1, x: "<abc>ABC</abc>"};

    // No error.
    io:println(j.fromJsonWithType(Foo)); // i=1 x=<abc>ABC</abc>
    // No error.
    io:println(j.fromJsonWithType(Bar)); // i=1 x=<abc>ABC</abc>

    // Then this should be ambiguous?
    io:println(j.fromJsonWithType(FooBar)); // i=1 x=<abc>ABC</abc>
}
```

Reply:
Shouldn't this be okay according to [this](https://github.com/ballerina-platform/ballerina-spec/issues/556) ?

Let's take this
```ballerina
 io:println(j.fromJsonWithType(FooBar));
```
When we reach L273, `convertibleTypes` will have only have `Bar`.  isn't it?

Reply:
Ah, didn't know of the "fromJsonWithType should only do something different from cloneWithType if cloneWithType would produce an error." part.
 
Yeah, this behaviour is correct then. Thanks!